### DISCUSSION TITLE
Slot-Wise Processing

### DISCUSSION THREAD

**Discussion by danielsfrt** [NVlabs/sionna#65] [Q&A] [answered]
Hello Sionna contributors,

I want to use Sionna to sequentially process slots with varying pilot symbols, similar to 5G NR DMRS pilots which differ depending on the current slot index. As a first step, I used the OFDMSystem class from [this tutorial](https://nvlabs.github.io/sionna/examples/Sionna_tutorial_part4.html#Benchmarking-the-Neural-Receiver) and simply extended its ResourceGrid to comprise a full frame of N slots.

However, if I would want to operate in a slot-wise manner instead, I would have to instantiate the now slot-sized ResourceGrid (with its specific pilots) and all affected layers of the processing chain every time that I call the model. This does not seem very convenient to me.

Do you have any recommendations how I could approach this issue?

Best regards,
Daniel

**Comment by jhoydis** [comment]
Hi Daniel,

I would need some more information to advise you on this:

- Does the pilot mask, i.e., the reserved resource elements, change from slot to slot?
- Does the number of pilots change from slot to slot?
- How many different pilot patterns would you need to create?

Thanks,

Jakob

**Reply by danielsfrt** [reply]
Hi Jakob,

Thank you for you for quick reponse!
* The pilot mask remains the same from slot to slot. Only the pilot symbols itself change as they are dependent on the slot index.
* For the moment, also the number of pilots would not change from slot to slot.
* I would need to create at least 10 different pilot patterns. My custom pilot pattern class inherits from ```sn.ofdm.PilotPattern``` and hence takes both pilot mask and symbols as parameters during instantiation.

Best regards,
Daniel

**Comment by jhoydis** [comment, marked as answer]
Hi Daniel,

I have quickly put together a working solution in this [notebook](https://colab.research.google.com/drive/11_o-SGAg5AbX_a3O1Tijq0QdHYGgqwzH?usp=sharing). However, this is not working in graph/xla mode due to the way the ``ResourceGridMapper`` is currently implemented.

In order to make it work in graph mode, you would need to somehow move (and slightly modify) the following lines from the ``build`` function of the [ResourceGridMapper](https://nvlabs.github.io/sionna/_modules/sionna/ofdm/resource_grid.html#ResourceGridMapper) to its ``call`` function:
```
pilots = flatten_last_dims(self._resource_grid.pilot_pattern.pilots, 3)
self._template = tf.scatter_nd(pilot_ind, pilots, rg_type.shape)
self._template = tf.expand_dims(self._template, -1)
```

Alternatively, you could instantiate for each pilot pattern a different ``ResourceGridMapper`` and store it in a list in the model. Depending on the slot index, which you would need to provide as an input to the model, you would select the correct mapper.

Hope that helps.
